## Ex:2 Data Wrangling and Transformation 
### Objective

To perform data wrangling and transformation on a dataset using Python and Pandas by handling missing values, removing duplicates, correcting data types, filtering data, and transforming variables into a suitable format for data analysis and machine learning.



##  Dataset Description

The dataset contains information about **student academic performance, educational background, MBA specialization, and placement salary**.

###  Dataset Attributes

| Column | Description |
|:---|:---|
| `sl_no` | Serial number / student identifier |
| `gender` | Gender of the student |
| `hsc_p` | Higher Secondary / 12th percentage |
| `hsc_s` | Higher Secondary stream |
| `degree_p` | Undergraduate degree percentage |
| `degree_t` | Undergraduate degree type |
| `etest_p` | Employability / entrance test percentage |
| `specialisation` | MBA specialization |
| `mba_p` | MBA percentage |
| `salary` | Salary offered after placement |

---

## Experiment Question

 **Using the given student placement dataset, perform data wrangling and transformation by handling missing values, scaling numerical features, detecting and treating outliers, encoding categorical variables, and generating a final model-ready dataset.**




Data wrangling and transformation is the process of converting raw student placement data into a clean, consistent, and machine-learning-ready dataset. In this experiment, the given student placement dataset is first loaded into a Pandas DataFrame and explored by examining its rows, columns, data types, and descriptive statistics. Missing values are then identified and handled by removing records with missing `salary` values and replacing missing values in `hsc_p`, `degree_p`, and `etest_p` with their respective mean values. The numerical attributes such as `hsc_p`, `degree_p`, `etest_p`, and `salary` are transformed using feature-scaling techniques such as `StandardScaler` and `MinMaxScaler` to bring the variables into suitable numerical ranges. The dataset is then divided into input features (`X`) and the target variable (`Y`), where `salary` is considered the target. The preprocessed data is saved as `Pre.csv` for further processing. Next, possible outliers in the `salary` attribute are identified using a boxplot and statistically detected using the Z-score method, where values with an absolute Z-score greater than 3 are considered potential outliers. Outliers are also treated using the capping and flooring method by calculating the 5th and 95th percentiles and replacing values outside these limits with the corresponding boundary values. The salary distribution before and after outlier treatment is then compared using visualization. Since the dataset also contains categorical attributes such as `gender`, `hsc_s`, `degree_t`, and `specialisation`, these variables are converted into numerical representations using `LabelEncoder` and One-Hot Encoding. Finally, the completely transformed dataset is verified for missing values, data types, dimensions, and numerical representation, and the resulting model-ready dataset is saved as `Final.csv`. Thus, the experiment demonstrates the complete workflow of preparing real-world tabular data for data analysis and machine-learning applications.

## Step 1: Start by importing the necessary Python libraries for data preprocessing.


In [4]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.preprocessing import LabelEncoder
from scipy.stats import zscore
from scipy import stats
from sklearn.preprocessing import LabelEncoder

In [2]:
import sys
print(sys.executable)

C:\Users\akhip\data_science_lab\Scripts\python.exe


## Step 2: Load the placement dataset into a Pandas Dataframe.

In [5]:
df = pd.read_csv("data.csv")
df

,sl_no,gender,hsc_p,hsc_s,degree_p,degree_t,etest_p,specialisation,mba_p,salary
0,1,M,91.00,Commerce,58.00,Sci&Tech,55.0,Mkt&HR,58.80,270000.0
1,2,M,78.33,Science,77.48,Sci&Tech,86.5,Mkt&Fin,66.28,200000.0
2,3,M,NaN,Arts,64.00,Comm&Mgmt,75.0,Mkt&Fin,57.80,250000.0
3,4,M,52.00,Science,NaN,Sci&Tech,66.0,Mkt&HR,59.43,NaN
4,5,M,73.60,Commerce,73.30,Comm&Mgmt,96.8,Mkt&Fin,55.50,425000.0
...,...,...,...,...,...,...,...,...,...,...
210,211,M,82.00,Commerce,77.60,Comm&Mgmt,91.0,Mkt&Fin,74.49,400000.0
211,212,M,60.00,Science,72.00,Sci&Tech,74.0,Mkt&Fin,53.62,275000.0
212,213,M,67.00,Commerce,73.00,Comm&Mgmt,59.0,Mkt&Fin,69.72,295000.0
213,214,F,66.00,Commerce,58.00,Comm&Mgmt,70.0,Mkt&HR,60.23,204000.0


## Step 3:Take a quick look at the data to understand its structure and identify any missing values or anomalies.

In [6]:
df.head()

,sl_no,gender,hsc_p,hsc_s,degree_p,degree_t,etest_p,specialisation,mba_p,salary
0,1,M,91.00,Commerce,58.00,Sci&Tech,55.0,Mkt&HR,58.80,270000.0
1,2,M,78.33,Science,77.48,Sci&Tech,86.5,Mkt&Fin,66.28,200000.0
2,3,M,NaN,Arts,64.00,Comm&Mgmt,75.0,Mkt&Fin,57.80,250000.0
3,4,M,52.00,Science,NaN,Sci&Tech,66.0,Mkt&HR,59.43,NaN
4,5,M,73.60,Commerce,73.30,Comm&Mgmt,96.8,Mkt&Fin,55.50,425000.0


In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 215 entries, 0 to 214
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sl_no           215 non-null    int64  
 1   gender          215 non-null    str    
 2   hsc_p           210 non-null    float64
 3   hsc_s           215 non-null    str    
 4   degree_p        213 non-null    float64
 5   degree_t        215 non-null    str    
 6   etest_p         211 non-null    float64
 7   specialisation  215 non-null    str    
 8   mba_p           214 non-null    float64
 9   salary          148 non-null    float64
dtypes: float64(5), int64(1), str(4)
memory usage: 16.9 KB


In [8]:
df.isnull().sum()

sl_no              0
gender             0
hsc_p              5
hsc_s              0
degree_p           2
degree_t           0
etest_p            4
specialisation     0
mba_p              1
salary            67
dtype: int64

In [9]:
df.dropna(subset=['salary'], inplace=True)

In [10]:
df.shape

(148, 10)

In [12]:
df['hsc_p'] = df['hsc_p'].fillna(df['hsc_p'].mean())

In [13]:
df['hsc_p'].isnull().sum()



np.int64(0)

In [14]:
df['degree_p'] = df['degree_p'].fillna(df['degree_p'].mean())

In [15]:
df['degree_p'].isnull().sum()

np.int64(0)

In [16]:
df['etest_p'] = df['etest_p'].fillna(df['etest_p'].mean())

In [17]:
df['etest_p'].isnull().sum()

np.int64(0)

#### The method isnull() checks each element in the DataFrame (or Series) to see if it is NaN (Not a Number) or None (missing value).
It returns a DataFrame (or Series) of the same shape as the input, with Boolean values:
#### True: The value is null (NaN or None).
#### False: The value is not null.

## Step 4: Handle Missing Data
### Option 1: If the dataset is large and only a small percentage of data is missing, you can remove rows with missing values using dropna(subset,inplace)


### Option 2:If removing data isn't ideal, you can impute (df.[""].fillna(df[""].mean(),inplace)) missing values using methods like mean, median, or most frequent.

In [18]:
df

,sl_no,gender,hsc_p,hsc_s,degree_p,degree_t,etest_p,specialisation,mba_p,salary
0,1,M,91.00000,Commerce,58.00,Sci&Tech,55.0,Mkt&HR,58.80,270000.0
1,2,M,78.33000,Science,77.48,Sci&Tech,86.5,Mkt&Fin,66.28,200000.0
2,3,M,70.00089,Arts,64.00,Comm&Mgmt,75.0,Mkt&Fin,57.80,250000.0
4,5,M,73.60000,Commerce,73.30,Comm&Mgmt,96.8,Mkt&Fin,55.50,425000.0
7,8,M,64.00000,Science,66.00,Sci&Tech,67.0,Mkt&Fin,62.14,252000.0
...,...,...,...,...,...,...,...,...,...,...
209,210,M,72.00000,Commerce,65.00,Comm&Mgmt,67.0,Mkt&Fin,56.49,216000.0
210,211,M,82.00000,Commerce,77.60,Comm&Mgmt,91.0,Mkt&Fin,74.49,400000.0
211,212,M,60.00000,Science,72.00,Sci&Tech,74.0,Mkt&Fin,53.62,275000.0
212,213,M,67.00000,Commerce,73.00,Comm&Mgmt,59.0,Mkt&Fin,69.72,295000.0


In [19]:
df.isnull().sum()

sl_no             0
gender            0
hsc_p             0
hsc_s             0
degree_p          0
degree_t          0
etest_p           0
specialisation    0
mba_p             0
salary            0
dtype: int64

## Step 5: Feature Scaling
Feature scaling is the process of converting numerical features to a similar scale so that one feature does not dominate another simply because it has larger numerical values.

<img src="https://i.postimg.cc/G21gMYnF/f.png" alt="Image Description" width="500">









## Option 1( StandardScaler): This method scales the data to have a mean of 0 and a standard deviation of 1.


In [20]:
scaler = StandardScaler()

In [21]:
df[['hsc_p', 'degree_p', 'etest_p', 'salary']] = scaler.fit_transform(
    df[['hsc_p', 'degree_p', 'etest_p', 'salary']]
)

In [22]:
df.head()

,sl_no,gender,hsc_p,hsc_s,degree_p,degree_t,etest_p,specialisation,mba_p,salary
0,1,M,2.265997e+00,Commerce,-1.652293,Sci&Tech,-1.328518,Mkt&HR,58.80,-0.200292
1,2,M,8.987875e-01,Science,1.346845,Sci&Tech,0.978332,Mkt&Fin,66.28,-0.951839
2,3,M,1.533482e-15,Arts,-0.728534,Comm&Mgmt,0.136149,Mkt&Fin,57.80,-0.415019
4,5,M,3.883770e-01,Commerce,0.703293,Comm&Mgmt,1.732635,Mkt&Fin,55.50,1.463849
7,8,M,-6.475513e-01,Science,-0.420614,Sci&Tech,-0.449718,Mkt&Fin,62.14,-0.393547


#### Option 2:This method scales the data to a fixed range, usually between 0 and 1. 
###  MinMaxScaler()

In [23]:
minmax = MinMaxScaler()

In [24]:
df[['hsc_p', 'degree_p', 'etest_p', 'salary']] = minmax.fit_transform(
    df[['hsc_p', 'degree_p', 'etest_p', 'salary']]
)

In [25]:
df.head()

,sl_no,gender,hsc_p,hsc_s,degree_p,degree_t,etest_p,specialisation,mba_p,salary
0,1,M,0.857051,Commerce,0.057143,Sci&Tech,0.104167,Mkt&HR,58.80,0.094595
1,2,M,0.586729,Science,0.613714,Sci&Tech,0.760417,Mkt&Fin,66.28,0.000000
2,3,M,0.409023,Arts,0.228571,Comm&Mgmt,0.520833,Mkt&Fin,57.80,0.067568
4,5,M,0.485812,Commerce,0.494286,Comm&Mgmt,0.975000,Mkt&Fin,55.50,0.304054
7,8,M,0.280990,Science,0.285714,Sci&Tech,0.354167,Mkt&Fin,62.14,0.070270


## Step 6  Option 1: Identifying Outliers Using Z-Scores
The value of 3 in the context of Z-scores is often used as a threshold to identify outliers in a dataset. A Z-score represents how many standard deviations a data point is away from the mean of the dataset. Specifically:

A Z-score of 0 means the data point is exactly at the mean.
A Z-score of 1 means the data point is one standard deviation above the mean, and so on.
A Z-score of 3 corresponds to a data point being 3 standard deviations away from the mean. For a normal distribution, about 99.7% of the data points fall within 3 standard deviations of the mean (according to the 68-95-99.7 rule, which describes the spread of data in a normal distribution). Therefore, points with Z-scores greater than 3 or less than -3 are considered unusually far from the mean and are often flagged as outliers.

This threshold (Z > 3 or Z < -3) is commonly used in many statistical applications because it captures the extreme values that are rare in a normal distribution, which are typically considered to be outliers. However, the choice of threshold can vary depending on the specific application and the nature of the data.



[![Chat-GPT-Image-Aug-20-2026-08-06-05-PM.png](https://i.postimg.cc/2SFTtcXL/Chat-GPT-Image-Aug-20-2026-08-06-05-PM.png)](https://postimg.cc/4Yyz75GX)

In [26]:
z_scores = zscore(df['salary'])

In [27]:
outliers = df[abs(z_scores) > 3]

In [28]:
outliers

,sl_no,gender,hsc_p,hsc_s,degree_p,degree_t,etest_p,specialisation,mba_p,salary
119,120,M,0.374867,Commerce,0.245714,Comm&Mgmt,0.482102,Mkt&Fin,64.34,1.000000
150,151,M,0.167058,Science,0.057143,Sci&Tech,0.125000,Mkt&Fin,61.30,0.662162
177,178,F,0.985065,Commerce,0.657143,Comm&Mgmt,0.812500,Mkt&Fin,70.81,0.608108


In [29]:
columns_to_check = ["salary"]

z_score = stats.zscore(df[columns_to_check])
u = (z_score > 3)
l = (z_score < -3)
index = u | l

clean_df = df[~index]

df.info()
clean_df.info()

<class 'pandas.DataFrame'>
Index: 148 entries, 0 to 213
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sl_no           148 non-null    int64  
 1   gender          148 non-null    str    
 2   hsc_p           148 non-null    float64
 3   hsc_s           148 non-null    str    
 4   degree_p        148 non-null    float64
 5   degree_t        148 non-null    str    
 6   etest_p         148 non-null    float64
 7   specialisation  148 non-null    str    
 8   mba_p           148 non-null    float64
 9   salary          148 non-null    float64
dtypes: float64(5), int64(1), str(4)
memory usage: 12.7 KB
<class 'pandas.DataFrame'>
Index: 145 entries, 0 to 213
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sl_no           145 non-null    int64  
 1   gender          145 non-null    str    
 2   hsc_p           145 non-null    float6

### Option 2:  Capping and Flooring Outliers
Capping and flooring is an outlier-treatment technique where extreme values are replaced with predefined boundary values instead of deleting the records.

In [30]:
l1 = df["salary"].quantile(0.05)
u1 = df["salary"].quantile(0.95)

l1, u1

(np.float64(0.008243243243243256), np.float64(0.30168918918918924))

In [31]:
df_capped = df.copy()

In [32]:
df_capped["salary"] = df_capped["salary"].clip(l1, u1)

In [33]:
df_capped.info()

<class 'pandas.DataFrame'>
Index: 148 entries, 0 to 213
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sl_no           148 non-null    int64  
 1   gender          148 non-null    str    
 2   hsc_p           148 non-null    float64
 3   hsc_s           148 non-null    str    
 4   degree_p        148 non-null    float64
 5   degree_t        148 non-null    str    
 6   etest_p         148 non-null    float64
 7   specialisation  148 non-null    str    
 8   mba_p           148 non-null    float64
 9   salary          148 non-null    float64
dtypes: float64(5), int64(1), str(4)
memory usage: 12.7 KB


In [34]:
df.head()

,sl_no,gender,hsc_p,hsc_s,degree_p,degree_t,etest_p,specialisation,mba_p,salary
0,1,M,0.857051,Commerce,0.057143,Sci&Tech,0.104167,Mkt&HR,58.80,0.094595
1,2,M,0.586729,Science,0.613714,Sci&Tech,0.760417,Mkt&Fin,66.28,0.000000
2,3,M,0.409023,Arts,0.228571,Comm&Mgmt,0.520833,Mkt&Fin,57.80,0.067568
4,5,M,0.485812,Commerce,0.494286,Comm&Mgmt,0.975000,Mkt&Fin,55.50,0.304054
7,8,M,0.280990,Science,0.285714,Sci&Tech,0.354167,Mkt&Fin,62.14,0.070270


## Step 7: Convert categorical variables into numerical format using LabelEncoder ().
[![Picture1.png](https://i.postimg.cc/yNpNvnVd/Picture1.png)](https://postimg.cc/zLW5fCHZ)




In [35]:
L1 = LabelEncoder()

df["gender"] = L1.fit_transform(df["gender"])
df["hsc_s"] = L1.fit_transform(df["hsc_s"])
df["degree_t"] = L1.fit_transform(df["degree_t"])
df["specialisation"] = L1.fit_transform(df["specialisation"])

df.head()

,sl_no,gender,hsc_p,hsc_s,degree_p,degree_t,etest_p,specialisation,mba_p,salary
0,1,1,0.857051,1,0.057143,2,0.104167,1,58.80,0.094595
1,2,1,0.586729,2,0.613714,2,0.760417,0,66.28,0.000000
2,3,1,0.409023,0,0.228571,0,0.520833,0,57.80,0.067568
4,5,1,0.485812,1,0.494286,0,0.975000,0,55.50,0.304054
7,8,1,0.280990,2,0.285714,2,0.354167,0,62.14,0.070270


## Convert categorical variables into numerical format using one hot encoder

[![Picture2.png](https://i.postimg.cc/gcZ0Hv1J/Picture2.png)](https://postimg.cc/HjTHp7YD)

In [36]:
c = ["gender"]

one_hot_encoded_data = pd.get_dummies(df, columns=c)

one_hot_encoded_data.head()

,sl_no,hsc_p,hsc_s,degree_p,degree_t,etest_p,specialisation,mba_p,salary,gender_0,gender_1
0,1,0.857051,1,0.057143,2,0.104167,1,58.80,0.094595,False,True
1,2,0.586729,2,0.613714,2,0.760417,0,66.28,0.000000,False,True
2,3,0.409023,0,0.228571,0,0.520833,0,57.80,0.067568,False,True
4,5,0.485812,1,0.494286,0,0.975000,0,55.50,0.304054,False,True
7,8,0.280990,2,0.285714,2,0.354167,0,62.14,0.070270,False,True


In [37]:
one_hot_encoded_data.to_csv('clean.csv', index=False)

In [38]:
Y = df["salary"].copy()

In [39]:
X = df.drop(columns=["salary", "sl_no"]).copy()

In [40]:
c = ["gender"]

one_hot_encoded_data = pd.get_dummies(df, columns=c)

one_hot_encoded_data.head()

,sl_no,hsc_p,hsc_s,degree_p,degree_t,etest_p,specialisation,mba_p,salary,gender_0,gender_1
0,1,0.857051,1,0.057143,2,0.104167,1,58.80,0.094595,False,True
1,2,0.586729,2,0.613714,2,0.760417,0,66.28,0.000000,False,True
2,3,0.409023,0,0.228571,0,0.520833,0,57.80,0.067568,False,True
4,5,0.485812,1,0.494286,0,0.975000,0,55.50,0.304054,False,True
7,8,0.280990,2,0.285714,2,0.354167,0,62.14,0.070270,False,True


In [41]:
one_hot_encoded_data.to_csv('clean.csv', index=False)

In [42]:
one_hot_encoded_data.head()

,sl_no,hsc_p,hsc_s,degree_p,degree_t,etest_p,specialisation,mba_p,salary,gender_0,gender_1
0,1,0.857051,1,0.057143,2,0.104167,1,58.80,0.094595,False,True
1,2,0.586729,2,0.613714,2,0.760417,0,66.28,0.000000,False,True
2,3,0.409023,0,0.228571,0,0.520833,0,57.80,0.067568,False,True
4,5,0.485812,1,0.494286,0,0.975000,0,55.50,0.304054,False,True
7,8,0.280990,2,0.285714,2,0.354167,0,62.14,0.070270,False,True


In [43]:
y = df['salary'].copy()

In [44]:
X = df.drop(columns=['salary', 'sl_no']).copy()

In [45]:
X.shape


(148, 8)

In [46]:
y.shape

(148,)

In [47]:
one_hot_encoded_data

,sl_no,hsc_p,hsc_s,degree_p,degree_t,etest_p,specialisation,mba_p,salary,gender_0,gender_1
0,1,0.857051,1,0.057143,2,0.104167,1,58.80,0.094595,False,True
1,2,0.586729,2,0.613714,2,0.760417,0,66.28,0.000000,False,True
2,3,0.409023,0,0.228571,0,0.520833,0,57.80,0.067568,False,True
4,5,0.485812,1,0.494286,0,0.975000,0,55.50,0.304054,False,True
7,8,0.280990,2,0.285714,2,0.354167,0,62.14,0.070270,False,True
...,...,...,...,...,...,...,...,...,...,...,...
209,210,0.451675,1,0.257143,0,0.354167,0,56.49,0.021622,False,True
210,211,0.665031,1,0.617143,0,0.854167,0,74.49,0.270270,False,True
211,212,0.195648,2,0.457143,2,0.500000,0,53.62,0.101351,False,True
212,213,0.344997,1,0.485714,0,0.187500,0,69.72,0.128378,False,True


In [48]:
one_hot_encoded_data.to_csv("Final.csv", index=False)

# Exercise: Data Cleaning and Transformation – Automobile Dataset

## Step 1: Load and Explore the Dataset

### 1. Load the Dataset
- Import Pandas and load the Automobile dataset.
- Display the first 10 rows.
- Display the shape of the dataset.

### 2. Explore the Dataset
- Display the column names.
- Display the data types.
- Generate descriptive statistics.
- Identify numerical and categorical columns.
- Display unique values in categorical columns.

## Step 2: Data Cleaning

### 3. Check Missing Values
- Check for missing values in each column.
- Display the number and percentage of missing values.

### 4. Handle Missing Values
- Replace missing numerical values using mean or median.
- Replace missing categorical values using mode.
- Verify that no missing values remain.

### 5. Remove Duplicate Records
- Check for duplicate rows.
- Display the number of duplicate records.
- Remove duplicate records.
- Verify the result.

### 6. Clean the `horsepower` Column
- Identify non-numeric values such as `?`.
- Replace `?` with `NaN`.
- Convert `horsepower` to numeric.
- Handle the resulting missing values.

## Step 3: Data Transformation

### 7. Transform the `origin` Column
- Display the unique values in `origin`.
- Convert the values into meaningful labels:
  - `1` → `usa`
  - `2` → `europe`
  - `3` → `japan`

### 8. Create `weight_kg`
- Create a new column `weight_kg`.
- Convert weight from pounds to kilograms.

  `weight_kg = weight × 0.453592`

### 9. Create `mpg_category`
Create a new column based on `mpg`:
- `< 20` → `Low`
- `20–29` → `Medium`
- `≥ 30` → `High`

### 10. Create `vehicle_age`
- Create a new column `vehicle_age`.
- Assume the current year is 2026.

  `vehicle_age = 2026 - model_year`

### 11. Rename Columns
Rename:
- `mpg` → `miles_per_gallon`
- `horsepower` → `hp`
- `weight` → `weight_lbs`
- `model_year` → `year`

### 12. Filter the Data
Display vehicles:
- With `mpg > 30`
- With `horsepower > 150`
- With `cylinders >= 6`
- Manufactured after 1980
- Originating from `usa`

## Step 4: Encoding Categorical Data

### 13. Label Encoding
- Apply `LabelEncoder` to the `origin` column.
- Create a new column `origin_encoded`.
- Display the original and encoded values.
- Display the category-to-label mapping.

### 14. One-Hot Encoding
- Apply One-Hot Encoding to the `origin` column.
- Compare Label Encoding and One-Hot Encoding.
- Which encoding method is more appropriate for `origin`? Explain why.

## Step 5: Outlier Detection

### 15. Identify Outliers Using Z-Scores
- Calculate the Z-score for the numerical features.
- Identify observations with `|Z-score| > 3` as outliers.
- Count the outliers in each numerical column.
- Display the rows containing outliers.
- Decide whether the outliers should be removed or retained.

## Step 6: Feature Scaling

### 16. Standardization Using StandardScaler
- Select the numerical features.
- Apply `StandardScaler`.
- Display the standardized values.
- Verify that the features have approximately mean `0` and standard deviation `1`.

## Step 7: Normalization

### 17. Normalization Using MinMaxScaler
- Apply `MinMaxScaler` to the numerical features.
- Transform the features to the range `[0, 1]`.
- Display the normalized values.
- Compare **Standardization** and **Normalization**.
- Explain when each scaling method is appropriate.

## Step 8: Create Features and Target

### 18. Create X and Y Variables
- Select the appropriate input features as **X (independent variables)**.
- Select `mpg` as **Y (target variable)**.
- Display the shape of `X` and `Y`.
- Save `X` and `Y` into `automobile_X_Y.csv`.
- Load the CSV file again and display the first 5 rows.

## Step 9: Save the Final Dataset

### 19. Save the Preprocessed Dataset
- Combine the processed features and target variable.
- Display the final dataset.
- Check for missing values.
- Save the final dataset as `automobile_preprocessed.csv`.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("Automobile.csv")

In [3]:
df.head(10)

,name,mpg,cylinders,displacement,horsepower,weight,acceleration,model_year,origin
0,chevrolet chevelle malibu,18.0,8.0,307.0,130.0,3504.0,12.0,70,usa
1,buick skylark 320,15.0,8.0,350.0,165.0,3693.0,11.5,70,usa
2,plymouth satellite,18.0,8.0,318.0,150.0,3436.0,11.0,70,usa
3,amc rebel sst,16.0,8.0,304.0,150.0,3433.0,12.0,70,usa
4,ford torino,17.0,NaN,302.0,140.0,3449.0,10.5,70,usa
5,ford galaxie 500,15.0,8.0,429.0,198.0,4341.0,10.0,70,usa
6,chevrolet impala,14.0,8.0,454.0,220.0,NaN,9.0,70,usa
7,plymouth fury iii,14.0,8.0,440.0,215.0,4312.0,8.5,70,usa
8,pontiac catalina,14.0,8.0,455.0,NaN,4425.0,10.0,70,usa
9,amc ambassador dpl,15.0,8.0,390.0,190.0,3850.0,8.5,70,usa


In [4]:
df.shape

(398, 9)

In [5]:
df.columns

Index(['name', 'mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'model_year', 'origin'],
      dtype='str')

In [6]:
df.dtypes

name                str
mpg             float64
cylinders       float64
displacement    float64
horsepower      float64
weight          float64
acceleration    float64
model_year        int64
origin              str
dtype: object

In [7]:
df.describe()

,mpg,cylinders,displacement,horsepower,weight,acceleration,model_year
count,398.000000,395.000000,395.000000,386.000000,396.000000,395.000000,398.000000
mean,23.514573,5.445570,193.340506,104.316062,2965.025253,15.562278,76.010050
std,7.815984,1.696203,104.425993,38.086281,845.254458,2.750260,3.697627
min,9.000000,3.000000,68.000000,46.000000,1613.000000,8.000000,70.000000
25%,17.500000,4.000000,102.500000,75.250000,2222.250000,13.850000,73.000000
50%,23.000000,4.000000,146.000000,92.500000,2797.500000,15.500000,76.000000
75%,29.000000,8.000000,262.000000,125.000000,3581.750000,17.150000,79.000000
max,46.600000,8.000000,455.000000,230.000000,5140.000000,24.800000,82.000000


In [8]:
df.select_dtypes(include='number').columns

Index(['mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'model_year'],
      dtype='str')

In [9]:
df["origin"].unique()

<StringArray>
['usa', 'japan', 'europe']
Length: 3, dtype: str

In [10]:
df["name"].unique()

<StringArray>
[ 'chevrolet chevelle malibu',          'buick skylark 320',
         'plymouth satellite',              'amc rebel sst',
                'ford torino',           'ford galaxie 500',
           'chevrolet impala',          'plymouth fury iii',
           'pontiac catalina',         'amc ambassador dpl',
 ...
 'chrysler lebaron medallion',             'ford granada l',
           'toyota celica gt',          'dodge charger 2.2',
           'chevrolet camaro',            'ford mustang gl',
                  'vw pickup',              'dodge rampage',
                'ford ranger',                 'chevy s-10']
Length: 305, dtype: str

In [11]:
df.isnull().sum()

name             0
mpg              0
cylinders        3
displacement     3
horsepower      12
weight           2
acceleration     3
model_year       0
origin           0
dtype: int64

In [12]:
(df.isnull().sum() / len(df)) * 100

name            0.000000
mpg             0.000000
cylinders       0.753769
displacement    0.753769
horsepower      3.015075
weight          0.502513
acceleration    0.753769
model_year      0.000000
origin          0.000000
dtype: float64

In [13]:
df["cylinders"] = df["cylinders"].fillna(df["cylinders"].mean())

In [14]:
df["cylinders"].isnull().sum()

np.int64(0)

In [15]:
df["displacement"] = df["displacement"].fillna(df["displacement"].mean())

In [16]:
df["displacement"].isnull().sum()

np.int64(0)

In [17]:
df["horsepower"] = df["horsepower"].fillna(df["horsepower"].mean())

In [18]:
df["horsepower"].isnull().sum()

np.int64(0)

In [19]:
df["weight"] = df["weight"].fillna(df["weight"].mean())

In [20]:
df["weight"].isnull().sum()

np.int64(0)

In [21]:
df["acceleration"] = df["acceleration"].fillna(df["acceleration"].mean())

In [22]:
df["acceleration"].isnull().sum()

np.int64(0)

In [23]:
df.isnull().sum()

name            0
mpg             0
cylinders       0
displacement    0
horsepower      0
weight          0
acceleration    0
model_year      0
origin          0
dtype: int64

In [24]:
df.duplicated().sum()

np.int64(0)

In [25]:
df = df.drop_duplicates()

In [26]:
df.duplicated().sum()

np.int64(0)

In [27]:
df["horsepower"].unique()

array([130.        , 165.        , 150.        , 140.        ,
       198.        , 220.        , 215.        , 104.31606218,
       190.        , 170.        , 160.        , 225.        ,
        95.        ,  97.        ,  85.        ,  88.        ,
        87.        ,  90.        , 113.        , 200.        ,
       210.        , 193.        , 100.        , 105.        ,
       175.        , 153.        , 180.        , 110.        ,
        72.        ,  86.        ,  70.        ,  76.        ,
        65.        ,  69.        ,  60.        ,  80.        ,
        54.        , 208.        , 155.        , 112.        ,
        92.        , 145.        , 137.        , 158.        ,
        46.        , 167.        ,  94.        , 107.        ,
       230.        ,  49.        ,  75.        ,  91.        ,
       122.        ,  67.        ,  83.        ,  78.        ,
        52.        ,  61.        ,  93.        , 148.        ,
       129.        ,  96.        ,  71.        ,  98.  

In [28]:
df["horsepower"].dtype

dtype('float64')

In [29]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

df["origin_encoded"] = le.fit_transform(df["origin"])

In [30]:
df[["origin", "origin_encoded"]].head()

,origin,origin_encoded
0,usa,2
1,usa,2
2,usa,2
3,usa,2
4,usa,2


In [31]:
dict(zip(le.classes_, le.transform(le.classes_)))

{'europe': np.int64(0), 'japan': np.int64(1), 'usa': np.int64(2)}

In [32]:
one_hot_origin = pd.get_dummies(df["origin"], prefix="origin")

In [33]:
one_hot_origin.head()

,origin_europe,origin_japan,origin_usa
0,False,False,True
1,False,False,True
2,False,False,True
3,False,False,True
4,False,False,True


In [34]:
df_onehot = pd.concat([df, one_hot_origin], axis=1)

In [35]:
df_onehot.head()

,name,mpg,cylinders,displacement,horsepower,weight,acceleration,model_year,origin,origin_encoded,origin_europe,origin_japan,origin_usa
0,chevrolet chevelle malibu,18.0,8.00000,307.0,130.0,3504.0,12.0,70,usa,2,False,False,True
1,buick skylark 320,15.0,8.00000,350.0,165.0,3693.0,11.5,70,usa,2,False,False,True
2,plymouth satellite,18.0,8.00000,318.0,150.0,3436.0,11.0,70,usa,2,False,False,True
3,amc rebel sst,16.0,8.00000,304.0,150.0,3433.0,12.0,70,usa,2,False,False,True
4,ford torino,17.0,5.44557,302.0,140.0,3449.0,10.5,70,usa,2,False,False,True


In [36]:
numeric_cols = df.select_dtypes(include="number").columns

In [37]:
numeric_cols

Index(['mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'model_year', 'origin_encoded'],
      dtype='str')

In [38]:
numeric_features = [
    "mpg",
    "cylinders",
    "displacement",
    "horsepower",
    "weight",
    "acceleration",
    "model_year"
]

In [39]:
numeric_features

['mpg',
 'cylinders',
 'displacement',
 'horsepower',
 'weight',
 'acceleration',
 'model_year']

In [40]:
from scipy import stats

In [41]:
z_scores = stats.zscore(df[numeric_features])

In [42]:
outlier_mask = (abs(z_scores) > 3)

In [43]:
outlier_mask.sum()

np.int64(6)

In [44]:
outlier_counts = pd.Series(
    outlier_mask.sum(axis=0),
    index=numeric_features
)

In [45]:
outlier_counts

mpg             0
cylinders       0
displacement    0
horsepower      4
weight          0
acceleration    2
model_year      0
dtype: int64

In [46]:
outlier_rows = df[outlier_mask.any(axis=1)]

In [47]:
outlier_rows

,name,mpg,cylinders,displacement,horsepower,weight,acceleration,model_year,origin,origin_encoded
6,chevrolet impala,14.0,8.0,454.0,220.0,2965.025253,9.0,70,usa,2
13,buick estate wagon (sw),14.0,8.0,455.0,225.0,3086.000000,10.0,70,usa,2
95,buick electra 225 custom,12.0,8.0,455.0,225.0,4951.000000,11.0,73,usa,2
116,pontiac grand prix,16.0,8.0,400.0,230.0,4278.000000,9.5,73,usa,2
299,peugeot 504,27.2,4.0,141.0,71.0,3190.000000,24.8,79,europe,0
394,vw pickup,44.0,4.0,97.0,52.0,2130.000000,24.6,82,europe,0


In [48]:
from sklearn.preprocessing import StandardScaler

In [49]:
scaler = StandardScaler()

In [50]:
scaled_data = scaler.fit_transform(df[numeric_features])

In [51]:
scaled_data

array([[-0.7064387 ,  1.51359459,  1.09393251, ...,  0.64006479,
        -1.30180992, -1.62742629],
       [-1.09075062,  1.51359459,  1.50779234, ...,  0.86451361,
        -1.48453144, -1.62742629],
       [-0.7064387 ,  1.51359459,  1.19980363, ...,  0.55931072,
        -1.66725297, -1.62742629],
       ...,
       [ 1.08701694, -0.85655353, -0.56150678, ..., -0.79569511,
        -1.44798714,  1.62198339],
       [ 0.57460104, -0.85655353, -0.70587649, ..., -0.40380035,
         1.11011422,  1.62198339],
       [ 0.95891297, -0.85655353, -0.71550114, ..., -0.29098216,
         1.40246866,  1.62198339]], shape=(398, 7))

In [52]:
scaled_df = pd.DataFrame(
    scaled_data,
    columns=numeric_features
)

In [53]:
scaled_df.head()

,mpg,cylinders,displacement,horsepower,weight,acceleration,model_year
0,-0.706439,1.513595,1.093933,0.685653,0.640065,-1.301810,-1.627426
1,-1.090751,1.513595,1.507792,1.620005,0.864514,-1.484531,-1.627426
2,-0.706439,1.513595,1.199804,1.219568,0.559311,-1.667253,-1.627426
3,-0.962647,1.513595,1.065059,1.219568,0.555748,-1.301810,-1.627426
4,-0.834543,0.000000,1.045809,0.952611,0.574749,-1.849974,-1.627426


In [54]:
scaled_df.mean()

mpg             7.141133e-17
cylinders       1.785283e-16
displacement   -8.926416e-17
horsepower     -2.856453e-16
weight          3.927623e-16
acceleration    6.605548e-16
model_year     -1.642461e-15
dtype: float64

In [55]:
scaled_df.std()

mpg             1.001259
cylinders       1.001259
displacement    1.001259
horsepower      1.001259
weight          1.001259
acceleration    1.001259
model_year      1.001259
dtype: float64

In [56]:
from sklearn.preprocessing import MinMaxScaler

In [57]:
minmax_scaler = MinMaxScaler()

In [58]:
normalized_data = minmax_scaler.fit_transform(df[numeric_features])

In [59]:
normalized_data

array([[0.2393617 , 1.        , 0.61757106, ..., 0.5361497 , 0.23809524,
        0.        ],
       [0.15957447, 1.        , 0.72868217, ..., 0.58973632, 0.20833333,
        0.        ],
       [0.2393617 , 1.        , 0.64599483, ..., 0.51686986, 0.17857143,
        0.        ],
       ...,
       [0.61170213, 0.2       , 0.17312661, ..., 0.19336547, 0.21428571,
        1.        ],
       [0.50531915, 0.2       , 0.13436693, ..., 0.2869294 , 0.63095238,
        1.        ],
       [0.58510638, 0.2       , 0.13178295, ..., 0.31386447, 0.67857143,
        1.        ]], shape=(398, 7))

In [60]:
normalized_df = pd.DataFrame(
    normalized_data,
    columns=numeric_features
)

In [61]:
normalized_df.head()

,mpg,cylinders,displacement,horsepower,weight,acceleration,model_year
0,0.239362,1.000000,0.617571,0.456522,0.536150,0.238095,0.0
1,0.159574,1.000000,0.728682,0.646739,0.589736,0.208333,0.0
2,0.239362,1.000000,0.645995,0.565217,0.516870,0.178571,0.0
3,0.186170,1.000000,0.609819,0.565217,0.516019,0.238095,0.0
4,0.212766,0.489114,0.604651,0.510870,0.520556,0.148810,0.0


In [62]:
Y = df["mpg"].copy()

In [63]:
Y.head()

0    18.0
1    15.0
2    18.0
3    16.0
4    17.0
Name: mpg, dtype: float64

In [64]:
X = df[
    ["cylinders",
     "displacement",
     "horsepower",
     "weight",
     "acceleration",
     "model_year",
     "origin_encoded"]
].copy()

In [65]:
X.head()

,cylinders,displacement,horsepower,weight,acceleration,model_year,origin_encoded
0,8.00000,307.0,130.0,3504.0,12.0,70,2
1,8.00000,350.0,165.0,3693.0,11.5,70,2
2,8.00000,318.0,150.0,3436.0,11.0,70,2
3,8.00000,304.0,150.0,3433.0,12.0,70,2
4,5.44557,302.0,140.0,3449.0,10.5,70,2


In [66]:
X.shape

(398, 7)

In [67]:
Y.shape

(398,)

In [68]:
automobile_X_Y = pd.concat([X, Y], axis=1)

In [69]:
automobile_X_Y.to_csv("automobile_X_Y.csv", index=False)

In [70]:
automobile_X_Y_loaded = pd.read_csv("automobile_X_Y.csv")

In [71]:
automobile_X_Y_loaded.head()

,cylinders,displacement,horsepower,weight,acceleration,model_year,origin_encoded,mpg
0,8.00000,307.0,130.0,3504.0,12.0,70,2,18.0
1,8.00000,350.0,165.0,3693.0,11.5,70,2,15.0
2,8.00000,318.0,150.0,3436.0,11.0,70,2,18.0
3,8.00000,304.0,150.0,3433.0,12.0,70,2,16.0
4,5.44557,302.0,140.0,3449.0,10.5,70,2,17.0


In [72]:
X_final = normalized_df.copy()

In [73]:
X_final.head()

,mpg,cylinders,displacement,horsepower,weight,acceleration,model_year
0,0.239362,1.000000,0.617571,0.456522,0.536150,0.238095,0.0
1,0.159574,1.000000,0.728682,0.646739,0.589736,0.208333,0.0
2,0.239362,1.000000,0.645995,0.565217,0.516870,0.178571,0.0
3,0.186170,1.000000,0.609819,0.565217,0.516019,0.238095,0.0
4,0.212766,0.489114,0.604651,0.510870,0.520556,0.148810,0.0


In [74]:
X_final = normalized_df.drop(columns=["mpg"]).copy()

In [75]:
X_final = pd.concat([X_final, one_hot_origin], axis=1)

In [76]:
X_final.head()

,cylinders,displacement,horsepower,weight,acceleration,model_year,origin_europe,origin_japan,origin_usa
0,1.000000,0.617571,0.456522,0.536150,0.238095,0.0,False,False,True
1,1.000000,0.728682,0.646739,0.589736,0.208333,0.0,False,False,True
2,1.000000,0.645995,0.565217,0.516870,0.178571,0.0,False,False,True
3,1.000000,0.609819,0.565217,0.516019,0.238095,0.0,False,False,True
4,0.489114,0.604651,0.510870,0.520556,0.148810,0.0,False,False,True


In [77]:
final_dataset = pd.concat([X_final, Y], axis=1)

In [78]:
final_dataset.head()

,cylinders,displacement,horsepower,weight,acceleration,model_year,origin_europe,origin_japan,origin_usa,mpg
0,1.000000,0.617571,0.456522,0.536150,0.238095,0.0,False,False,True,18.0
1,1.000000,0.728682,0.646739,0.589736,0.208333,0.0,False,False,True,15.0
2,1.000000,0.645995,0.565217,0.516870,0.178571,0.0,False,False,True,18.0
3,1.000000,0.609819,0.565217,0.516019,0.238095,0.0,False,False,True,16.0
4,0.489114,0.604651,0.510870,0.520556,0.148810,0.0,False,False,True,17.0


In [79]:
final_dataset.isnull().sum()

cylinders        0
displacement     0
horsepower       0
weight           0
acceleration     0
model_year       0
origin_europe    0
origin_japan     0
origin_usa       0
mpg              0
dtype: int64

In [80]:
final_dataset.to_csv("automobile_preprocessed.csv", index=False)

In [81]:
final_dataset.head()

,cylinders,displacement,horsepower,weight,acceleration,model_year,origin_europe,origin_japan,origin_usa,mpg
0,1.000000,0.617571,0.456522,0.536150,0.238095,0.0,False,False,True,18.0
1,1.000000,0.728682,0.646739,0.589736,0.208333,0.0,False,False,True,15.0
2,1.000000,0.645995,0.565217,0.516870,0.178571,0.0,False,False,True,18.0
3,1.000000,0.609819,0.565217,0.516019,0.238095,0.0,False,False,True,16.0
4,0.489114,0.604651,0.510870,0.520556,0.148810,0.0,False,False,True,17.0
